File: 2025-09-September-GFG-POTD/September 24 - Design MinMax Queue/Solution.py
from collections import deque

class SpecialQueue:
    def __init__(self):
        self.q = deque()
        self.minDeque = deque()
        self.maxDeque = deque()

    def enqueue(self, x: int) -> None:
        self.q.append(x)

        # Maintain minDeque in increasing order
        while self.minDeque and self.minDeque[-1] > x:
            self.minDeque.pop()
        self.minDeque.append(x)

        # Maintain maxDeque in decreasing order
        while self.maxDeque and self.maxDeque[-1] < x:
            self.maxDeque.pop()
        self.maxDeque.append(x)

    def dequeue(self) -> None:
        if self.q:
            val = self.q.popleft()

            if self.minDeque and self.minDeque[0] == val:
                self.minDeque.popleft()
            if self.maxDeque and self.maxDeque[0] == val:
                self.maxDeque.popleft()

    def getFront(self) -> int:
        return self.q[0]

    def getMin(self) -> int:
        return self.minDeque[0]

    def getMax(self) -> int:
        return self.maxDeque[0]

File: 2025-09-September-GFG-POTD/September 24 - Design MinMax Queue/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class SpecialQueue {
private:
    queue<int> q;
    deque<int> minDeque;
    deque<int> maxDeque;

public:
    SpecialQueue() {}

    void enqueue(int x) {
        q.push(x);

        // Maintain minDeque in increasing order
        while (!minDeque.empty() && minDeque.back() > x) {
            minDeque.pop_back();
        }
        minDeque.push_back(x);

        // Maintain maxDeque in decreasing order
        while (!maxDeque.empty() && maxDeque.back() < x) {
            maxDeque.pop_back();
        }
        maxDeque.push_back(x);
    }

    void dequeue() {
        if (!q.empty()) {
            int val = q.front();
            q.pop();

            if (!minDeque.empty() && minDeque.front() == val) {
                minDeque.pop_front();
            }
            if (!maxDeque.empty() && maxDeque.front() == val) {
                maxDeque.pop_front();
            }
        }
    }

    int getFront() {
        return q.front();
    }

    int getMin() {
        return minDeque.front();
    }

    int getMax() {
        return maxDeque.front();
    }
};
